File: populatingNextRightPointers.py
#Time Complexity : O(n) where n is the number of nodes in the tree
#Space Complexity : O(1)
#Did this code successfully run on Leetcode : Yes

"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution:
    #BFS
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        l = root
        while l.left:
            curr = l
            while curr:
                curr.left.next = curr.right
                if curr.next:
                    curr.right.next = curr.next.left
                curr = curr.next
            l = l.left
        return root

    #DFS - TC -> O(n) SC -> O(h)
    # def helper(self, l, r):
    #     if not l:
    #         return
    #     l.next = r
    #     self.helper(l.left, l.right)
    #     self.helper(l.right, r.left)
    #     self.helper(r.left, r.right)
    #
    # def connect(self, root: 'Node') -> 'Node':
    #     if not root:
    #         return
    #     self.helper(root.left, root.right)
    #     return root


#BRUTE FORCE - TC -> O(n) SC -> O(n)
#         if root is None:
#             return None
#         queue = [root]
#         while queue:
#             l = len(queue)
#             for i in range(l):
#                 curr = queue.pop(0)
#                 if i < l-1:
#                     curr.next = queue[0]
#                 if curr.left:
#                     queue.append(curr.left)
#                 if curr.right:
#                     queue.append(curr.right)

#         return root

File: populating_next_right_pointers.cpp
// Time Complexity : O(n)
// Space Complexity :  O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this 

class Solution {
private:
    vector<vector<Node*>> vect;
    
    void level_order_trav(Node* root)
    {
        if(root == NULL)
            return;
        
        queue<Node*> nodeq;
        queue<int> levelq;
        
        Node* temp;
        int tlevel;
        
        nodeq.push(root);
        levelq.push(0);
        
        // Keep pushing in Queue and access sequentially
        while(!nodeq.empty())
        {
            temp = nodeq.front();
            nodeq.pop();
            
            tlevel = levelq.front();
            levelq.pop();
            
            cout << temp->val << endl;

            // If same level, add there   
            if(vect.size() == tlevel)
            {
                vector<Node*> temp_vect;
                
                temp_vect.push_back(temp);
                
                vect.push_back(temp_vect);
            }
            // If not, create new level
            else
                vect[tlevel].push_back(temp);
            
            // Push left node
            if(temp->left)
            {
                nodeq.push(temp->left);
                levelq.push(tlevel+1);
            }
            
            // Push Right node
            if(temp->right)
            {
                nodeq.push(temp->right);
                levelq.push(tlevel+1);
            }
        }  
    }
    
public:
    Node* connect(Node* root) {
        
        // Do level order traversal
        // And put the node in vector 
        level_order_trav(root);
        
        int j;
        
        for(int i = 0; i < vect.size(); i++)
        {
            j = 0;
            // Update the next pointers
            while(j < vect[i].size()-1)
            {
                vect[i][j]->next = vect[i][j+1];
                j++;
            }
            
        }
        return root;
    }
};